[V1][Frontend] Improve Shutdown And Logs

rafvasq opened the pull request:
In place of #11737

authored by: @robertgshaw2-redhat 

SUMMARY:

- Prior to this PR, if we encountered an error in a background process, we kill the process tree immediately, which means that we cannot cleanup resources and cannot return good status codes to clients. This PR overhauls the Error handling to instead shut down the background processes and raise Errors that allow us to return proper HTTP status codes to users
- Prior to this PR, we were not properly shutting down when Errors occured during startup, especially in the TP case
- Prior to this PR, we used signals to catch errors from background processes. Due to limitations of Python, this prevented us from running outside the main thread. This is a problem for deployments in TritonServer

DESIGN:

- for errors during startup, we wrap __init__ code with try...catch and push FAILED over the ready PIPE. This works well since the parent processes are waiting for confirmation
- for errors during runtime, we wrap the busy loops with try..catch and push failure messages over the existing IPC mechanisms.

- One weakness is that issues with the ipc mechanisms themselves are not handled explicitly. We will explore this more in a follow up

TEST MATRIX:

- AsyncLLM, TP=1 + TP>1 --- runtime and startup
- LLM (MP), TP=1, TP>1 --- runtime and startup
- LLM (no-MP), TP=1, TP>1 --- runtime and startup

Fixes: https://github.com/vllm-project/vllm/issues/12690

Commit:
checkpoint prototype

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
Issue currently is with streaming. The HTTP exception handlers do not handle properly

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
switch from ValueError -> Exception.

Commit:
merged

Commit:
updated

Commit:
stash

Commit:
stash

Commit:
add watchdog

Commit:
updated

Commit:
revert spurious changes

Commit:
updated

Commit:
updated

Commit:
updated

Commit:
remove cruft

Commit:
cruft

Commit:
stash

Commit:
fix llama

Commit:
updated

Commit:
cruft

Commit:
cruft

Commit:
updated

Commit:
updated

Commit:
updated

Commit:
updated

Commit:
update comment

Commit:
update comment

Commit:
fix more

Commit:
updated

Commit:
udpatd

Commit:
added exception file

Commit:
updated

Commit:
fixt

Commit:
reduce cruft

Commit:
reduce cruft

Commit:
cleanup

Commit:
updated

Commit:
cruft

Commit:
updated

Commit:
revert changes to server

Commit:
revert debug cruft

Commit:
fix error

Commit:
added tests

Commit:
revert

Commit:
fixed

Commit:
updated

Commit:
fixed error

Commit:
update test coverage

Commit:
stash

Commit:
added tests

Commit:
stash

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updared

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
fix typo

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
stash

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
remove signal handler

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
remove signal handler

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
update comment

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
avoid sigusr1

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
cleanup

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
cleanup

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
cleanup

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
cleanup

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
it starts?

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
nits

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
fix test for bunched streaming

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
tweak typing

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

robertgshaw2-redhat committed:
Update tests/v1/shutdown/test_forward_error.py

Co-authored-by: Cyrus Leung <cyrus.tl.leung@gmail.com>

robertgshaw2-redhat committed:
Merge branch 'main' into api-server-error-handling

Commit:
pre commit

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

robertgshaw2-redhat committed:
Update tests/v1/shutdown/test_forward_error.py

Co-authored-by: Russell Bryant <rbryant@redhat.com>

robertgshaw2-redhat committed:
Update vllm/v1/engine/core.py

Co-authored-by: Russell Bryant <rbryant@redhat.com>

robertgshaw2-redhat committed:
Update vllm/v1/engine/core.py

Co-authored-by: Russell Bryant <rbryant@redhat.com>

robertgshaw2-redhat committed:
Update tests/v1/shutdown/test_forward_error.py

Co-authored-by: Russell Bryant <rbryant@redhat.com>

rafvasq committed:
Rebase

Signed-off-by: Rafael Vasquez <rafvasq21@gmail.com>

rafvasq committed:
Merge branch 'main' of https://github.com/vllm-project/vllm into api-server-error-handling

robertgshaw2-redhat commented:
Can you also address the comments on https://github.com/vllm-project/vllm/pull/11737?

rafvasq committed:
Small fixes, splits startup tests to sim OOM

Signed-off-by: Rafael Vasquez <rafvasq21@gmail.com>

rafvasq committed:
Lints

Signed-off-by: Rafael Vasquez <rafvasq21@gmail.com>

rafvasq commented:
@robertgshaw2-redhat I addressed all of the comments except for a couple open-ended conversations:
- https://github.com/vllm-project/vllm/pull/11737#discussion_r1949875674
- https://github.com/vllm-project/vllm/pull/11737#discussion_r1949885640

Commit:
cleanup

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
cleaning up

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
cleanup pr

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
cleanup pr

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updatd

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
update to ensure we call shutdown on RPC error

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
fixed

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
updated

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
removed mp=0 tests

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
removed non mp tests

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
removed non mp tests

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

Commit:
fixed error

Signed-off-by: rshaw@neuralmagic.com <rshaw@neuralmagic.com>

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -240,8 +243,24 @@ async def _generate(
+            await self.abort(request_id)
+            if self.log_requests:
+                logger.info("Request %s failed.", request_id)
+            logger.exception("GOT EXCEPTION:", exc_info=e)
```
Is this line meant to be here? We should probably change the message if so, and there's no need to set `exc_info=e` since it will use that by default in the except block.

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -240,8 +243,24 @@ async def _generate(
+        except Exception as e:
+            await self.abort(request_id)
+            if self.log_requests:
+                logger.info("Request %s failed.", request_id)
```
Maybe add the exception message to the log line too? (but still single line)

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -410,16 +432,20 @@ async def pin_lora(self, lora_id: int) -> bool:
     def is_running(self) -> bool:
-        return True
+        # Have not started the loop yet.
+        if self.output_handler is None:
+            return True
+
+        return not self.output_handler.done()
```
```suggestion
        return self.output_handler is None or not self.output_handler.done()
```

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -410,16 +432,20 @@ async def pin_lora(self, lora_id: int) -> bool:
     @property
     def errored(self) -> bool:
-        return False
+        return (self.engine_core.is_engine_dead or not self.is_running)
```
nit:
```suggestion
        return self.engine_core.is_engine_dead or not self.is_running
```

njhill commented on vllm/v1/engine/core.py:
```diff
@@ -247,25 +249,37 @@ def __init__(
+            # Send Readiness signal to EngineClient.
+            ready_pipe.send({"status": "READY"})
+
+        except Exception as e:
+            logger.exception("EngineCore got an Exception during startup:",
+                             exc_info=e)
+            ready_pipe.send({"status": "FAILED"})
+            raise e
+
+        finally:
+            ready_pipe.close()
```
I'm not sure that this should go here. Probably better to catch/handle in the calling method, including handling sending the pipe ready message since it "owns" the pipe. In the failure case, I'm not sure that we need to send failed status back since the process could die for other reasons without us being able to do so, and we need to handle that equivalently from the client side.

njhill commented on vllm/v1/executor/multiproc_executor.py:
```diff
@@ -250,35 +275,56 @@ def make_worker_process(
 
-        worker_response_mq = MessageQueue.create_from_handle(
-            worker_response_mq_handle, 0)
+    @staticmethod
+    def wait_for_ready(
+            unready_proc_handle: UnreadyWorkerProcHandle) -> WorkerProcHandle:
```
why not a method on `UnreadyProcHandle`?

njhill reviewed (commented):
I'd like to spend more time on this but left comments from a first pass.

rafvasq commented:
@robertgshaw2-redhat, can I help out with a refactor of this one again?

rafvasq commented:
Back to https://github.com/vllm-project/vllm/pull/11737
